### PR TITLE
ScriptContext (and transitive deps) types

### PR DESCRIPTION
**Pull request by t1lde**
Addressing #43 Draft because I haven't tested this yet,

Derived with the help of some TH, but the end result needed a lot of manual tweaking, so the TH isn't included.

It was a bit of a headache reasoning about wrapping with PAsData vs not, and how to correctly derive based on that.

The assumption I went with is that values of `PCurrencySymbol` are unwrapped, which `PasData PCurrency` symbol would be wrapped as Data. 
This lines up with the return type of `pdhead`, too, but definitely has some gotcha potential.

Further work is to be done adding API functions such as field accessors.

### CODE REVIEW

**Review by L-as — CHANGES_REQUESTED** [review]
You probably want to add support for the `pconstant` stuff.

**Comment by L-as** [comment]
I want to note that if you mess up the `PAsData` stuff, you will just get type checking errors when you try to `pdhead` it, i.e. it's not unsafe if you make a mistake in that specific regard, nor is it a breaking change to fix it later.

File: Plutarch/Api/V2.hs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE DerivingVia #-}
+module Plutarch.Api.V2
```
[the remark this answers is not in the document]
Comment (t1lde) [reply Plutarch/Api/V2.hs:2]: I have been under the impression that the V2 ScriptContext is always been supported on the ledger, with (compatible) differences in de-serialization...

I really hope that's the case, too. `txInfoRedeemers` is pretty necessary.

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Api/V2.hs:2]: `txInfoRedemeers` is not real. It's never been real. It's essentially just an elaborate prank. https://github.com/input-output-hk/plutus/issues/4285

[the remark this answers is not in the document]
Comment (t1lde) [reply Plutarch/Api/V2.hs:2]: Ouch.

**Comment by L-as** [comment]
FYI there are still merge conflicts

File: Plutarch/Lift.hs [hunk 2]
```diff
@@ -107,3 +112,28 @@ instance
 
 showEvalException :: EvaluationException CekUserError (MachineError PLC.DefaultFun) (UPLC.Term UPLC.DeBruijn PLC.DefaultUni PLC.DefaultFun ()) -> Text
 showEvalException = T.pack . show
+
+{- | 
+  DerivingVia wrapper for deriving `PLift` instances
+  via the wrapped type, while lifting to a coercible Haskell type.
+
+-}
+newtype PLiftVia (p :: k -> Type) (h :: Type) s = PLiftVia (p s)
```
Comment (L-as) [inline Plutarch/Lift.hs:121]: Don't you want `(s :: k)`?

**Review by L-as — APPROVED** [review]
LGTM

File: Plutarch/Api/V1.hs [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+  = PTxInfo
+      ( Term
+          s
+          ( PDataList
+              '[ PBuiltinList PTxInInfo
+               , PBuiltinList PTxOut
+               , PValue
+               , PValue
+               , PBuiltinList PDCert
+               , PBuiltinList (PBuiltinPair PStakingCredential PInteger)
```
[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Api/V1.hs:91]: Looks like it's actually a builtin list of `Constr` data values. The Haskell tuple here has a `ToData` instance that creates a `Constr 0 [Data]`.

I think the `PIsData (PBuiltinPair PInteger (PBuiltinList PData))` should be removed in favor of having `(PIsData a, PIsData b)=> PIsData (PBuiltinPair a b)`

File: Plutarch/Api/V1.hs [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+               , PPOSIXTimeRange
+               , PBuiltinList PPubKeyHash
```
Comment (TotallyNotChase) [inline Plutarch/Api/V1.hs:93]: Curious question: Shouldn't this be `PBuiltinList (PAsData PPubKeyHash)` - `PPubKeyHash` is not a data value, rather a builtin bytestring - right? So its data representation would be `PAsData PPubKeyHash`? Compared to, say `PTxInfo`, which *is* a data value the entire time - not a builtin that would be converted to a data value in this context.

File: Plutarch/Api/V1.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Api/V1.hs:91]: The second instance wouldn't be valid AFAICT?

[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Api/V1.hs:91]: Oh right, the `pfromData` wouldn't be possible. But the relationship seems to be `PlutusTx.toData (a, b) == Constr 0 [toData a, toData b]` and `PlutusTx.fromData (Constr 0 [toData a, toData b]) == Just (a, b)`.

File: Plutarch/Builtin.hs [hunk 5]
```diff
@@ -71,6 +76,20 @@ ptailBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
...
+  Note that the Plutus API doesn't use this most of the time,
+  instead encoding as a List of Tuple constructors.
+
+  Not to be confused with `PlutusTx.AssocMap.Map` / `PMap`
+
+-}
+type PBuiltinMap a b = (PBuiltinList (PBuiltinPair (PAsData a) (PAsData b)))
+
+-- | List type used for Plutus `Data`'s List constructor.
+type PList a  = PBuiltinList (PAsData a)
```
Comment (TotallyNotChase) [inline Plutarch/Builtin.hs:91]: I worry about whether or not it's redundant to write something like `PAsData PTxInfo` (in the case of `PList PTxInfo`). After all `PTxInfo` *is* `Data`, as opposed to, say `PPubKeyHash` - which does require this treatment.

Comment (L-as) [reply Plutarch/Builtin.hs:91]: The internal representation of `TxInfo` is an implementation detail.

Comment (t1lde) [reply Plutarch/Builtin.hs:91]: There's a choice I've made here for consistency -
all of the types are assumed to be 'unwrapped' by default from the Data constructor. 

As @L-as says, the internal repr is an implementation detail, the repr of `PAsData` can be arbitrarily.

**Comment by t1lde** [comment]
I think this is ready, just need to get out of build hell to check the last merges...

File: Plutarch/DataRepr.hs [hunk 6]
```diff
@@ -136,3 +146,36 @@ instance (Ledger.FromData h, Ledger.ToData h, PIsData p) => PLift (PIsDataReprIn
     maybeToRight "Failed to decode data" $ Ledger.fromData h
     where
       maybeToRight e = maybe (Left e) Right
+
+{- | 
+  DerivingVia wrapper for deriving `PLift` instances
+  via the wrapped type, while lifting to a coercible Haskell type.
+
+-}
+newtype PVia (p :: k -> Type) (h :: Type) (s :: k) = PVia (Term s p)
```
Comment (srid) [inline Plutarch/DataRepr.hs:155]: If this is used to derive only `PLift` instances, I suggest putting it in `Plutarch.Lift` module, alongside `PBuiltinType` - because it looks like they serve, essentially, the same purpose (with the only difference being that `PVia` is used to derive new types, and `PBuiltinType` is used to derive builtin types).

I also wonder why it is named `PVia` (rather than say `PNewType` to be consistent with `PBuiltinType` naming).

Comment (t1lde) [reply Plutarch/DataRepr.hs:163]: Good point, yeah. I'll move it over.

I'd say `PNewType` is less accurate than `PVia`, since it uses `Coercible`, it's a bit more powerful than just newtypes (same deal with deriving via vs deriving newtype, really).

Comment (t1lde) [reply Plutarch/DataRepr.hs:163]: Actually, I'll keep `PVia` where it is, to avoid the module cycle due to the `PlutusType` instance.

File: Plutarch/List.hs [hunk 7]
```diff
@@ -151,6 +152,19 @@ plength = phoistAcyclic $
     )
     $ \go -> plam $ \xs -> go # xs # 0
 
+{- |
+  Unsafely index a BuiltinList,
+  throwing an error if the index is out of bounds.
+-}
+punsafeIndex :: (PIsListLike list a) => Term s (PInteger :--> list a :--> a)
```
Comment (L-as) [inline Plutarch/List.hs:159]: This isn't unsafe FWIW.

File: plutarch.cabal [hunk 8]
```diff
@@ -1,39 +1,32 @@
...
-
+author:             Las Safin <me@las.rs>
+license:            MIT
 extra-source-files: README.md
 
 flag development
   description: Enable tracing functions within plutarch.
-  manual: True
-  default: False
+  manual:      True
```
Comment (L-as) [inline plutarch.cabal:10]: Why did you change all the formatting here?

**Review by L-as — APPROVED** [review]
Looks mostly good to me, but I think we should merge this PR so other people can build on top of it.

I don't have many complaints wrt. the code but the PR is too big IMO. The lifting stuff could have been done in a separate PR, and would have meant we could have processed the work incrementally.

**Comment by L-as** [comment]
#96

File: plutarch.cabal (hunk 8 above)
Comment (t1lde) [reply plutarch.cabal:10]: cabal-fmt did it
